module/zfs/dmu_send.c

```diff
@@ -2135,15 +2418,25 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
 	 * If we are losing blkptrs or changing the block size this must
 	 * be a new file instance.  We must clear out the previous file
 	 * contents before we can change this type of metadata in the dnode.
+	 * Raw receives will also check that the indirect structure of the
+	 * dnode hasn't changed.
 	 */
 	if (err == 0) {
-		int nblkptr;
-
-		nblkptr = deduce_nblkptr(drro->drr_bonustype,
+		uint32_t indblksz = drro->drr_indblkshift ?
+		    1ULL << drro->drr_indblkshift : 0;
+		int nblkptr = deduce_nblkptr(drro->drr_bonustype,
 		    drro->drr_bonuslen);
 
+		/* nblkptr will be bounded by the bonus size and type */
+		if (DRR_IS_RAW_ENCRYPTED(drro->drr_flags) &&
+		    nblkptr != drro->drr_nblkptr)
```

Comment:
Yes, in case our assumptions ever change in the future. It is important that we have that number correct for encrypted datasets so we can maintain the objset authentication structure.